src/plonk/lookup/prover.rs

```diff
@@ -0,0 +1,652 @@
+            ))
+            // z'(X) (a'(X) + \beta) (s'(X) + \gamma)
+            // - z'(\omega^{-1} X) (\theta^m a_1(X) + \theta^{m-1} a_2(X) + ... + a_m(X) + \beta) (\theta^m s_1(X) + \theta^{m-1} s_2(X) + ... + s_m(X) + \gamma)
+            .chain({
+                // z'(X) (a'(X) + \beta) (s'(X) + \gamma)
+                let mut left = self.product_coset.clone();
+                parallelize(&mut left, |left, start| {
+                    for ((left, permuted_input), permuted_table) in left
+                        .iter_mut()
+                        .zip(permuted.permuted_input_coset[start..].iter())
+                        .zip(permuted.permuted_table_coset[start..].iter())
+                    {
+                        *left *= &(*permuted_input + &(*beta));
+                        *left *= &(*permuted_table + &(*gamma));
+                    }
+                });
+
+                //  z'(\omega^{-1} X) (\theta^m a_1(X) + \theta^{m-1} a_2(X) + ... + a_m(X) + \beta) (\theta^m s_1(X) + \theta^{m-1} s_2(X) + ... + s_m(X) + \gamma)
+                let mut right = self.product_inv_coset;
+                let mut input_terms = pk.vk.domain.empty_extended();
+
+                // Compress the unpermuted input columns
+                for input in permuted.unpermuted_input_cosets.iter() {
+                    // \theta^m a_1(X) + \theta^{m-1} a_2(X) + ... + a_m(X)
+                    parallelize(&mut input_terms, |input_term, start| {
+                        for (input_term, input) in input_term.iter_mut().zip(input[start..].iter())
+                        {
+                            *input_term *= &(*theta);
+                            *input_term += input;
+                        }
+                    });
+                }
+
+                let mut table_terms = pk.vk.domain.empty_extended();
+                // Compress the unpermuted table columns
+                for table in permuted.unpermuted_table_cosets.iter() {
+                    //  \theta^m s_1(X) + \theta^{m-1} s_2(X) + ... + s_m(X)
+                    parallelize(&mut table_terms, |table_term, start| {
+                        for (table_term, table) in table_term.iter_mut().zip(table[start..].iter())
+                        {
+                            *table_term *= &(*theta);
+                            *table_term += table;
+                        }
+                    });
+                }
+
+                // Add \beta and \gamma offsets
+                parallelize(&mut right, |right, start| {
+                    for ((right, input_term), table_term) in right
+                        .iter_mut()
+                        .zip(input_terms[start..].iter())
+                        .zip(table_terms[start..].iter())
+                    {
+                        *right *= &(*input_term + &(*beta));
+                        *right *= &(*table_term + &(*gamma));
+                    }
+                });
```

Comment:
Similar to above, we can combine these loops.

```suggestion
                parallelize(&mut right, |right, start| {
                    for (i, right) in right.iter_mut().enumerate() {
                        let i = i + start;

                        // Compress the unpermuted input columns
                        let mut input_term = C::Scalar::zero();
                        for input in permuted.unpermuted_input_cosets.iter() {
                            input_term *= &theta;
                            input_term += &input[i];
                        }

                        // Compress the unpermuted table columns
                        let mut table_term = C::Scalar::zero();
                        for table in permuted.unpermuted_table_cosets.iter() {
                            table_term *= &theta;
                            table_term += &table[i];
                        }

                        // Add \beta and \gamma offsets
                        *right *= &(input_term + &beta);
                        *right *= &(table_term + &gamma);
                    }
                });
```